compiler/rustc_codegen_llvm/src/builder.rs

```diff
@@ -24,17 +25,255 @@ use rustc_span::Span;
 use rustc_target::callconv::FnAbi;
 use rustc_target::spec::{HasTargetSpec, SanitizerSet, Target};
 use smallvec::SmallVec;
-use tracing::{debug, instrument};
+use tracing::{debug, instrument, trace};
 
 use crate::abi::FnAbiLlvmExt;
 use crate::attributes;
 use crate::common::Funclet;
 use crate::context::CodegenCx;
-use crate::llvm::{self, AtomicOrdering, AtomicRmwBinOp, BasicBlock, False, True};
+use crate::llvm::AttributePlace::Function;
+use crate::llvm::{self, AtomicOrdering, AtomicRmwBinOp, BasicBlock, False, Metadata, True};
 use crate::type_::Type;
 use crate::type_of::LayoutLlvmExt;
 use crate::value::Value;
 
+fn get_params(fnc: &Value) -> Vec<&Value> {
+    unsafe {
+        let param_num = llvm::LLVMCountParams(fnc) as usize;
+        let mut fnc_args: Vec<&Value> = vec![];
+        fnc_args.reserve(param_num);
+        llvm::LLVMGetParams(fnc, fnc_args.as_mut_ptr());
+        fnc_args.set_len(param_num);
+        fnc_args
+    }
+}
+
+// The lowering of one `#[autodiff]` macro happens in multiple steps.
+// First we transalte generate a new dummy function, who's llvm-ir we now have as outer_fn.
+// We kept track of the original function to which the `#[autodiff]` macro was applied to, which we
+// now have as fn_to_diff. In our current implementation, we use the enzyme pass to carry out the
+// differentiation, following naming and calling conventions documented here: <https://enzyme.mit.edu/getting_started/CallingConvention/>
+//
+// Our `outer_fn` had some dummy code inserted at higher levels, so we first remove most of the
+// existing body. We then insert an `__enzyme_<autodiff/fwddiff>_<unique_id>` call, which the pass
+// will then pick up. FIXME(ZuseZ4): We will later want to upstream safety checks to the `outer_fn`,
+// in order to cover some assumptions of enzyme/autodiff, which could lead to UB otherwise.
```

Comment:
```suggestion
/// When differentiating `fn_to_diff`, take a `outer_fn` and generate another
/// function with expected naming and calling conventions[^1] which will be
/// discovered by the enzyme LLVM pass and its body populated with the differentiated
/// `fn_to_diff`. `outer_fn` is then modified to have a call to the generated
/// function and handle the differences between the Rust calling convention and
/// Enzyme.
/// [^1]: <https://enzyme.mit.edu/getting_started/CallingConvention/> 
// FIXME(ZuseZ4): `outer_fn` should include upstream safety checks to
// cover some assumptions of enzyme/autodiff, which could lead to UB otherwise.
```
Is this understanding correct?